x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDatafeedAction.java

```diff
@@ -248,6 +271,10 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeVLong(startTime);
             out.writeOptionalLong(endTime);
             out.writeVLong(timeout.millis());
+            if (out.getVersion().onOrAfter(Version.CURRENT)) {
+                out.writeOptionalWriteable(datafeedConfig);
+                out.writeOptionalWriteable(job);
+            }
         }
 
         @Override
```

Comment:
It seems wrong that `toXContent()` persists different information to `writeTo()`.  This means that the behaviour on full cluster restart is different to behaviour within a running cluster.  If someone did a full cluster restart with started datafeeds I suspect that things wouldn't work on restart.  (It should be possible to add a test in `x-pack/qa/full-cluster-restart` to confirm this.)

However, I think the solution to this problem is not to put the whole datafeed config in the X-Content representation.  That creates the same problem we have now where search syntax that's been removed prevents reading the cluster state from disk.

This leads to the conclusion that instead of storing the whole datafeed config in the task params we should store just enough fields from it to validate allocation, and definitely not the search or aggregations.  Then `toXContent()` can be consistent with `writeTo()`.

Reply:
The migration process will only move datafeeds from cluster state to index when the entire cluster has been upgraded to whatever version this goes into (6.6 or 6.7).  So if we can find the minimum node version in the cluster in `toXContent()` then we can write the extra fields into the X-Content representation only after the entire cluster has been upgraded.  That will make full cluster restarts work in the case where the entire cluster is on 6.6/6.7 (and it is essential this works because some people will run 6.7 for a year after 7.0 is released).

So if `job_id` is `null` after a full cluster restart then that implies the cluster was not completely upgraded to 6.6/6.7, and hence the migration will not have started, and hence the information can be obtained from the `MlMetadata` in cluster state.

Reply:
The reasoning is this:

- The job id is required to find the job task related to the datafeed and hence make the allocation decision
- If job_id is null then the task was not started from a migrated config/upgraded node. i.e. it is a running task from a full cluster restart or rolling upgrade
- In that case the datafeed config should be present in the clusterstate and the job_id can be found that way 
- It is not possible to modify the persistent task parameters 
- The persistent task may migrate once again before the task is stopped due to node failure and once again job_id will have to be found via the datafeed config in the clusterstate
- Therefore config cannot be removed from the clusterstate until the task has stopped
- Migration should duplicate the config in an index document and remove it from clusterstate on datafeed stopping

That seems reasonable enough until we hit a datafeed job created in 6.4 which is not stopped before 8

> So if we can find the minimum node version in the cluster in toXContent() ... 

I can't see a way of doing this and it won't help with full cluster restarts from 6.5

I could add the code to read job_id from the datafeed config in the clusterstate in this PR but I have no mechanism of testing it yet and have not started on the migration code. My preference is to finish the work on making jobs run from the new config and make it stable then tackle this issue (and others) in the migration work.